### PR TITLE
[IPC HLE] Implement /dev/usb/ven with very basic functionality

### PR DESCRIPTION
**Pull request by mbc07**
A little backstory: IOS 58 was introduced with the game Your Shape and later on with Wii System Menu 4.3. It's the only "official" way of accessing USB devices on the Wii at USB 2.0 speeds and it's exposed to the games/homebrew apps through /dev/usb/ven when that IOS is loaded. 

Newer Just Dance games (JD2014 and onwards) allow using an USB microphone to sing in the songs but sadly they'll always freeze at boot in Dolphin. The old wiispeak branch doesn't exhibit this issue, and after some digging into Dolphin logs, the last thing those games do before completely freezing is opening /dev/usb/ven, probably to check if a compatible USB microphone is available (no surprises, Ubisoft games trying to access something introduced earlier by another Ubisoft game and that's guaranteed to be present on any console updated to Wii System Menu 4.3).

So, here comes this PR. Although I didn't fully understand how IPC HLE works, I managed to isolate the code responsible for handling /dev/usb/ven on the old wiispeak branch (this PR is hugely based on it, so kudos to degasus/godisgovernment) and reintroduced it now in Dolphin with minimal functionality (basically correctly "answering" the game queries on /dev/usb/ven by saying there's no USB devices attached and nothing more). And guess what? That's enough to make the affected games happy.

TL;DR this PR fixes freeze at boot in Just Dance 2014, 2015, 2016 and probably in the upcoming Just Dance 2017 as well (yes, newer JD games are still released nowadays for the Wii, and it's very very likely they'll be reusing the same engine introduced in JD2014 again)

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4142)

### CODE REVIEW

**Comment by lioncash** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp, line 73 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpWCwOOsV70uklEARp:-KPpWCwPm0OAqzXZst-E:btpb6pz) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp#L73)):*
> ```cpp
>       // sent on change
>       static bool firstcall = true;
> 	  if (firstcall)
> ```

There's a tab here instead of spaces.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp, line 118 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpWG6bbNktBXuTu-bW:-KPpWG6c6UZ2nXS-WXNG:bwppkdz) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp#L118)):*
> ```cpp
> u32 CWII_IPC_HLE_Device_usb_ven::Update()
> {
> 	return IWII_IPC_HLE_Device::Update();
> ```

there's a tab here when it should be a space.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp, line 121 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpWK16pAClxSLZVOTE:-KPpWK16pAClxSLZVOTF:b-idxpt6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp#L121)):*
> ```cpp
> }
> 
> void CWII_IPC_HLE_Device_usb_ven::DoState(PointerWrap &p)
> ```

ampersand against the type.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.h, line 15 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpVfY8O005RNOEZvRj:-KPpVfY9gnOTL2gQi1yI:bi346p4) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.h#L15)):*
> ```objc
>   CWII_IPC_HLE_Device_usb_ven(u32 DeviceID, const std::string& DeviceName);
> 
>   virtual ~CWII_IPC_HLE_Device_usb_ven();
> ```

Virtual functions can just utilize the `override` keyword instead of `virtual` in this case. e.g.

```cpp
~CWII_IPC_HLE_Device_usb_ven() override;
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142)*

**Comment by JMC47** [comment]
This sounds really fun.  I'll be testing it later.

**Comment by mbc07** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp, line 73 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpWCwOOsV70uklEARp:-KPpZKuHzSvgzqqrmcYc:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp#L73)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> There's a tab here instead of spaces.
</details>
Done.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp, line 118 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpWG6bbNktBXuTu-bW:-KPpZNxV2pDx36ptF1EB:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp#L118)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> there's a tab here when it should be a space.
</details>
Done.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp, line 121 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpWK16pAClxSLZVOTE:-KPpZc5o3FS_daBH3Fiy:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.cpp#L121)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ampersand against the type.
</details>
Done.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.h, line 15 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-KPpVfY8O005RNOEZvRj:-KPpZdof0ftmBDfT0QRW:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/db2566c1e64173e1546a70d6262176b719953f2a/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_ven.h#L15)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Virtual functions can just utilize the `override` keyword instead of `virtual` in this case. e.g.
> 
> ```cpp
> ~CWII_IPC_HLE_Device_usb_ven() override;
> ```
> 
> 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142)*

**Comment by mbc07** [comment]
Just saw the lint log. This new commit should make it happy...

**Comment by JMC47** [comment]
Did some testing: all my USB Mic games are working much better.  Yes, they don't get in-game because there's no mic, but I think that's better than going in-game with no mic input or crashing outright.

LGTM

**Comment by mbc07** [comment]
Ping

**Comment by phire** [comment]
:lgtm:

---

Reviewed 4 of 6 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4142#-:-KR1IZJzZc-xHbMKEE-6:bnfp4nl)*
